documentation/src/main/asciidoc/mapper-orm-bridge-index-field-dsl.asciidoc

```diff
@@ -165,3 +165,105 @@ and pass it to the bridge for later use.
 <6> Create sub-fields, get references to these fields
 and pass them to the bridge for later use.
 ====
+
+[[mapper-orm-bridge-index-field-dsl-dynamic]]
+== Dynamic fields with field templates
+
+Field declared in the sections above are all _static_:
+their path and type are known on bootstrap.
+
+In some very specific cases, the path of a field is not known until you actually index it;
+for example, you may want to index a `Map<String, Integer>` by using the map keys as field names.
+The field, then, is considered _dynamic_.
+
+Dynamic fields are not declared on bootstrap,
+but need to match a field _template_ that is declared on bootstrap.
+The template includes the field type and structural information (multi-valued or not, ...),
+but omits the field name.
+
+Field templates are declared with the `fieldTemplate` methods, as shown below.
```

Comment:
Reading the chapter I think that it is not clear which interface the following `bind` method belongs to.
Maybe we should write something like:
> The template is activated on a give property using a property binding annotation.

Moreover, ideally I think that we should use a different annotation to bind dynamic fields.

Reply:
> Reading the chapter I think that it is not clear which interface the following bind method belongs to.

Right. I added something to clarify that.

> Moreover, ideally I think that we should use a different annotation to bind dynamic fields.

In order to do that, we would need one annotation per type: `@FullTextFieldTemplate`, `@KeywordFieldTemplate`, ... Which is a bit annoying.

But more importantly, dynamic fields are... well, dynamic. The annotation-based mapping is intrinsically static: all the fields are declared from the start. Typically, they are not useful for POJOs that are mapped "directly" with annotations.

They become useful when you need to map, say, a `Map<String, ?>`, or a `JsonObject`. Then you can't declare the fields upfront, but you may be able to define templates. However, when you do that, it's likely that you will need some custom code. Maybe the maps contain other maps as values and you need some recursion; maybe you don't want to index *everything* in the map.

My point is, you will need custom code (a bridge) at some point, so users who need templates already have a class ready to declare them. With that in mind, annotations feel a bit pointless, in my opinion.

I suppose we could introduce annotations for very specific use cases in the future. `@DynamicJsonFields` would be an idea. But in any case, that doesn't make the current API obsolete, as it will still be needed when we implement support for `@DynamicJsonFields` (for example).